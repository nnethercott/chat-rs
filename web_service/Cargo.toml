[package]
name = "web_service"
version = "0.1.0"
edition = "2024"

[lib]
path = "./src/lib.rs"

[[bin]]
path = "./src/main.rs"
name = "web"

[dependencies]
config = { version = "0.15.11", default-features = false, features=['yaml'] }
tokio = { version = "1.44.2", features = ["macros", "rt-multi-thread", "signal"] }
serde = { version = "1.0.219", features=["derive"] }
axum = "0.8.3"
anyhow = "1.0.98"
tonic = "0.13.0"
thiserror = "2.0.12"
grpc_service = {path="../grpc_service/"}
futures = "0.3.31"
