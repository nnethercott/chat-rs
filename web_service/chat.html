<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8">
        <title>WebSocket Chat</title>
    </head>
    <body>
        <h1>chat-rs</h1>

        <textarea id="chat" style="display:block; width:600px; height:400px; box-sizing: border-box" cols="30" rows="10"></textarea>
        <input id="input" style="display:block; width:600px; box-sizing: border-box" type="text" placeholder="chat">

        <script>
            const textarea = document.querySelector("#chat");
            const input = document.querySelector("#input");

           this.disabled = true;

            // establish connection with our websocket
            // TODO: establish connection only on `chat` 
            const websocket = new WebSocket("ws://localhost:3000/models/1/chat");

            websocket.onopen = function() {
                console.log("connection opened");
            }

            websocket.onclose = function() {
                console.log("connection closed");
            }

            websocket.onmessage = function(e) {
                console.log("agent: "+e.data);
                textarea.value += e.data;
            }

            input.onkeydown = function(e) {
                if (e.key == "Enter") {
                    textarea.value += "user: " + input.value + "\r\nagent: ";
                    websocket.send(input.value);
                    input.value = "";
                }
            }
        </script>
    </body>
</html>
