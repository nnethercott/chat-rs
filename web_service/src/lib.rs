pub mod config;
pub mod errors;
pub mod routes;

use std::{ops::Deref, sync::Arc};

use crate::config::get_config;
use axum::{
    Json, Router,
    extract::State,
    response::{IntoResponse, Response},
    routing::get,
};
use errors::WebError;
use futures::StreamExt;
use grpc_service::inferencer_client::InferencerClient;
use tokio::{net::TcpListener, sync::Mutex};
use tonic::transport::Channel;

async fn hello() -> &'static str {
    "hello, world"
}

async fn list_models(State(client): State<AppState>) -> Result<Json<Vec<String>>, WebError> {
    let stream = client
        .lock()
        .await
        .list_models(())
        .await
        .map_err(WebError::GrpcError)?
        .into_inner();

    let models: Vec<String> = stream
        .filter_map(|i| async { i.ok() })
        .map(|spec| spec.model_id)
        .collect()
        .await;

    Ok(Json(models))
}

#[derive(Clone)]
pub(crate) struct AppState(Arc<Mutex<InferencerClient<Channel>>>);

impl Deref for AppState{
    type Target = Mutex<InferencerClient<Channel>>;
    fn deref(&self) -> &Self::Target {
        &self.0.deref()
    }
}

pub async fn run_app() -> anyhow::Result<()> {
    let config = get_config()?;

    let grpc_client = InferencerClient::connect(format!("http://{}", config.grpc.addr())).await?;
    let state = AppState(Arc::new(Mutex::new(grpc_client)));

    let app = Router::new()
        .route("/", get(hello))
        .route("/models", get(list_models))
        .with_state(state);

    let listener = TcpListener::bind(config.server.addr()).await.unwrap();
    axum::serve(listener, app).await.unwrap();
    Ok(())
}

// gonna store these in a redis db per user session id
// #[derive(Serialize, Deserialize)]
// struct ChatHistory {
//     messages: Vec<Message>,
// }
//
// #[derive(Serialize, Deserialize)]
// struct Message {
//     role: Role,
//     content: String,
// }
//
// #[derive(Serialize, Deserialize)]
// enum Role {
//     User,
//     Assistant,
// }
