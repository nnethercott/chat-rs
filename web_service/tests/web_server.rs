use axum::{
    Router,
    body::Body,
    http::{self, Request, StatusCode},
};
use http_body_util::BodyExt;
use insta::{assert_debug_snapshot, assert_snapshot};
use tower::ServiceExt;
use web_service::{
    config::Settings,
    server::{App, AppState},
};

// a test not using axum-test
#[tokio::test]
async fn test_health() {
    let app = App::new(Settings::default()).unwrap().into_router();

    let response = app
        .with_state(AppState::default())
        .oneshot(
            Request::builder()
                .method(http::Method::GET)
                .uri("/health")
                .body(Body::empty())
                .unwrap(),
        )
        .await
        .unwrap();

    assert_eq!(response.status(), StatusCode::OK);

    // empty body but adding this as a reference for later ...
    let bytes = response.into_body().collect().await.unwrap().to_bytes();
    assert_debug_snapshot!(bytes);
}
