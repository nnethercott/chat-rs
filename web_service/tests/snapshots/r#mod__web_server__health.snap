---
source: web_service/tests/web_server.rs
expression: bytes
---
b""
